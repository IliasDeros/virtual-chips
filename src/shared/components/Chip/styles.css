.chip {
  @apply relative inline-block align-middle;
  --border-size: 2px;
  margin-top: -var(--border-size);
  width: var(--chip-size);
  height: var(--chip-size);
  border-radius: 100%;
  border: var(--border-size) white dotted;
  box-shadow: 2px 2px 0 rgba(0, 0, 0, 0.1);
}

.chip,
.chip--10 {
  background-color: crimson;
}
.chip--1 {
  background-color: #444;
}
.chip--5 {
  background-color: goldenrod;
}

.chip--x + .chip--x {
  left: -10px;
  margin-right: -10px;
}

/* prettier-ignore */
.chip--y { --index-stack: 0; }
/* prettier-ignore */
.chip--y + .chip--y:nth-child(2) { --index-stack: 1; }
/* prettier-ignore */
.chip--y + .chip--y:nth-child(3) { --index-stack: 2; }
/* prettier-ignore */
.chip--y + .chip--y:nth-child(4) { --index-stack: 3; }
/* prettier-ignore */
.chip--y + .chip--y:nth-child(5) { --index-stack: 4; }
/* prettier-ignore */
.chip--y + .chip--y:nth-child(6) { --index-stack: 5; }
/* prettier-ignore */
.chip--y + .chip--y:nth-child(7) { --index-stack: 6; }
/* prettier-ignore */
.chip--y + .chip--y:nth-child(8) { --index-stack: 7; }
/* prettier-ignore */
.chip--y + .chip--y:nth-child(9) { --index-stack: 8; }
/* prettier-ignore */
.chip--y + .chip--y:nth-child(10) { --index-stack: 9; }
/* prettier-ignore */
.chip--y + .chip--y:nth-child(11) { --index-stack: 10; }
/* prettier-ignore */
.chip--y + .chip--y:nth-child(12) { --index-stack: 11; }
/* prettier-ignore */
.chip--y + .chip--y:nth-child(13) { --index-stack: 12; }
/* prettier-ignore */
.chip--y + .chip--y:nth-child(14) { --index-stack: 13; }
/* prettier-ignore */
.chip--y + .chip--y:nth-child(14) { --index-stack: 13; }
/* prettier-ignore */
.chip--y + .chip--y:nth-child(15) { --index-stack: 14; }
/* prettier-ignore */
.chip--y + .chip--y:nth-child(16) { --index-stack: 15; }
/* prettier-ignore */
.chip--y + .chip--y:nth-child(17) { --index-stack: 16; }
/* prettier-ignore */
.chip--y + .chip--y:nth-child(18) { --index-stack: 17; }
/* prettier-ignore */
.chip--y + .chip--y:nth-child(19) { --index-stack: 18; }
/* prettier-ignore */
.chip--y + .chip--y:nth-child(20) { --index-stack: 19; }

/* Do not support vertical stacks of over 20 chips */
.chip--y + .chip--y:nth-child(20) ~ .chip--y {
  display: none;
}

.chip--y + .chip--y {
  --chip-size-neg: calc(var(--chip-size) * -1);
  left: var(--chip-size-neg);
  margin-right: var(--chip-size-neg);
  margin-top: calc(-10px * var(--index-stack));
}
