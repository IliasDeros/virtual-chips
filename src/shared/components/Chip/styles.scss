/*
 * CSS-rendered chips
 * Reference: https://codepen.io/Rovak/pen/ExYeQar
 * Large-Scale Reference: https://codepen.io/yakovd33/pen/jbbbPM
 * Denominations Reference: https://automaticpoker.com/poker-basics/poker-chip-values-for-home-game/
 *
 * Contents
 * # Chip Values
 * # Horizontal Stacking
 * # Vertical Stacking
 * # General Stacking
 */

:root {
  --chip-size: 1rem;
  --chip-size-neg: calc(var(--chip-size) * -1);
}

/************************************************
 # Chip Values
 ************************************************/
.chip {
  @apply relative inline-block align-middle;
  --border-size: 2px;
  margin-top: -var(--border-size);
  width: var(--chip-size);
  height: var(--chip-size);
  border-radius: 100%;
  border: var(--border-size) white dotted;
  box-shadow: 2px 2px 0 rgba(0, 0, 0, 0.1);
}

.chip,
.chip--10 {
  background-color: crimson;
}
.chip--1 {
  background-color: #444;
}
.chip--5 {
  background-color: goldenrod;
}

/************************************************
 # Horizontal Stacking
 ************************************************/
.chip--x + .chip--x {
  left: -10px;
  margin-right: -10px;
}

/************************************************
 # Vertical Stacking
 ************************************************/
$max-vertical-chips: 20;

.chip--y {
  --index-y: 0;
}

@for $i from 2 through $max-vertical-chips {
  .chip--y + .chip--y:nth-child(#{$i}) {
    --index-y: #{$i - 1};
  }
}

/* Do not support vertical stacks of over 20 chips */
.chip--y + .chip--y:nth-child(#{$max-vertical-chips}) ~ .chip--y {
  display: none;
}

.chip--y + .chip--y {
  left: var(--chip-size-neg);
  margin-right: var(--chip-size-neg);
  margin-top: calc(-10px * var(--index-y));
}

/************************************************
 # General Stacking
 ************************************************/
.chip--stack {
  $max-vertical-chips: 6;
  $stack: 0;
  $max-stacks: 8;
  $last-child-index: $max-stacks * $max-vertical-chips;

  @for $s from 0 through ($max-stacks - 1) {
    @for $y from 1 through ($max-vertical-chips - 1) {
      $child-index: $max-vertical-chips * $s + $y;
      $next-child-index: $child-index + 1;

      &:nth-child(#{$child-index})
        + .chip--stack:nth-child(#{$next-child-index}) {
        left: var(--chip-size-neg);
        margin-right: var(--chip-size-neg);
        margin-top: calc(-10px * #{$y});
      }
    }
  }

  // Hide extra chips
  &:nth-child(#{$last-child-index}) ~ .chip--stack {
    display: none;
  }
}
