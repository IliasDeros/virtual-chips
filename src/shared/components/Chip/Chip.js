import "./styles.scss";

export const Chip = ({
  className = "",
  stack = false,
  stackHorizontal = false,
  stackVertical = false,
  value = "10",
  ...htmlProps
}) => {
  let fullClass = `${className} chip chip--${value}`;

  if (stack) {
    fullClass += " chip--stack";
  }

  if (stackHorizontal) {
    fullClass += " chip--x";
  }

  if (stackVertical) {
    fullClass += " chip--y";
  }

  return <div className={fullClass} {...htmlProps}></div>;
};
