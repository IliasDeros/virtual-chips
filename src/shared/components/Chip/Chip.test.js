import { Chip } from "./Chip";
import { render } from "@testing-library/react";

it("Renders default chip", () => {
  const { container } = render(<Chip />);
  expect(container.firstChild).toMatchSnapshot();
});

it("Renders chip with custom value", () => {
  const { container } = render(<Chip value="5" />);
  expect(container.firstChild).toMatchSnapshot();
});

it("Renders stacked chip", () => {
  const { container } = render(<Chip stack />);
  expect(container.firstChild).toMatchSnapshot();
});

it("Renders horizontally stacked chip", () => {
  const { container } = render(<Chip stackHorizontal />);
  expect(container.firstChild).toMatchSnapshot();
});

it("Renders vertically stacked chip", () => {
  const { container } = render(<Chip stackVertical />);
  expect(container.firstChild).toMatchSnapshot();
});

it("Renders custom class", () => {
  const { container } = render(<Chip className="custom-class" />);
  expect(container.firstChild).toMatchSnapshot();
});
