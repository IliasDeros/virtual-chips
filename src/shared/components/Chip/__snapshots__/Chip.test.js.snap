// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders chip with custom value 1`] = `
<div
  class="chip chip--5"
/>
`;

exports[`Renders custom class 1`] = `
<div
  class="custom-class chip chip--10"
/>
`;

exports[`Renders default chip 1`] = `
<div
  class="chip chip--10"
/>
`;

exports[`Renders horizontally stacked chip 1`] = `
<div
  class="chip chip--10 chip--x"
/>
`;

exports[`Renders stacked chip 1`] = `
<div
  class="chip chip--10 chip--stack"
/>
`;

exports[`Renders vertically stacked chip 1`] = `
<div
  class="chip chip--10 chip--y"
/>
`;
